cpp/src/binaryop/compiled/struct_binary_ops.cuh

```diff
@@ -0,0 +1,136 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "binary_ops.hpp"
+#include "operation.cuh"
+
+#include <cudf/binaryop.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/table/experimental/row_operators.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::binops::compiled::detail {
+template <class T, class... Ts>
+inline constexpr bool is_any_v = std::disjunction<std::is_same<T, Ts>...>::value;
+
+template <typename DeviceComparator, class AdaptableUnaryFunction, class Iterator>
+struct device_comparison_functor {
+  using optional_iterator = thrust::transform_iterator<AdaptableUnaryFunction, Iterator>;
+  device_comparison_functor(optional_iterator const& optional_iter,
+                            bool const is_lhs_scalar,
+                            bool const is_rhs_scalar,
+                            DeviceComparator const& comparator)
+    : _optional_iter(optional_iter),
+      _is_lhs_scalar(is_lhs_scalar),
+      _is_rhs_scalar(is_rhs_scalar),
+      _comparator(comparator){};
+
+  bool __device__ operator()(size_type i)
+  {
+    return _optional_iter[i].has_value() &&
+           _comparator(cudf::experimental::row::lhs_index_type{_is_lhs_scalar ? 0 : i},
+                       cudf::experimental::row::rhs_index_type{_is_rhs_scalar ? 0 : i});
+  }
+
+  optional_iterator const _optional_iter;
+  bool const _is_lhs_scalar;
+  bool const _is_rhs_scalar;
+  DeviceComparator const _comparator;
+};
+
+template <class BinaryOperator,
+          typename PhysicalElementComparator =
+            cudf::experimental::row::lexicographic::sorting_physical_element_comparator>
+void apply_struct_binary_op(mutable_column_view& out,
+                            column_view const& lhs,
+                            column_view const& rhs,
+                            bool is_lhs_scalar,
+                            bool is_rhs_scalar,
+                            PhysicalElementComparator comparator = {},
+                            rmm::cuda_stream_view stream         = cudf::default_stream_value)
+{
+  auto compare_orders   = std::vector<order>(lhs.size(),
+                                           is_any_v<BinaryOperator, ops::Greater, ops::GreaterEqual>
+                                               ? order::DESCENDING
+                                               : order::ASCENDING);
+  auto tlhs             = table_view{{lhs}};
+  auto trhs             = table_view{{rhs}};
+  auto table_comparator = cudf::experimental::row::lexicographic::two_table_comparator{
+    tlhs, trhs, compare_orders, {}, stream};
+  auto outd = column_device_view::create(out, stream);
+  auto optional_iter =
+    cudf::detail::make_optional_iterator<bool>(*outd, nullate::DYNAMIC{out.has_nulls()});
+  auto nullate = nullate::DYNAMIC{has_nested_nulls(tlhs) || has_nested_nulls(trhs)};
```

Comment:
Seems like this name is shadowing the `cudf::nullate` enum class. We usually try to avoid this by choosing another name. (Note: another name to avoid is `has_nulls`, which shadows [`cudf::has_nulls`](https://docs.rapids.ai/api/libcudf/nightly/namespacecudf.html#a0b69e95220c2cc259bbf510506b6fff9).
```suggestion
  auto const comparator_nulls = nullate::DYNAMIC{has_nested_nulls(tlhs) || has_nested_nulls(trhs)};
```